skbio/sequence/_grammared_sequence.py

```diff
@@ -809,12 +808,21 @@ def to_definites(self, degenerate="wild", noncanonical=True):
                     f"Invalid character '{degenerate}' in sequence. Character must "
                     f"be within sequence alphabet: {self.alphabet}"
                 )
+        elif isinstance(degenerate, str) and len(degenerate) == 0:
+            sub_char = 0
         else:
             raise ValueError('Invalid value for parameter "degenerate".')
 
-        sequence = "".join(sub_char if x in degenerates else x for x in sequence)
+        htab = np.zeros((256,), dtype=np.uint8)
+        htab[self._degenerate_codes] = 1
+        if noncanonical:
+            htab[self._noncanonical_codes] = 1
+        pos = htab[self._bytes]
+        if sub_char:
```

Comment:
Here, I would suggest separating the situations of 1) trimming off degenerate characters and 2) replacing them with a new character. For 2), you don't have to do `np.trim_zeros`, which can save some compute.

For 1), there are a few problems: a) `np.trim_zeros` only removing leading and trailing zeros, but not middle ones. b) 0 can be a valid character code (although it is not usually used). We can't just make 0 a symbol for removal. I would suggest this instead: `self._bytes[np.where(1 - pos)[0]]`. (Also, I would suggest changing the parameter back to `trim` because it is more explicit than an empty string.)

Reply:
PS: I am thinking that "del" may be better than "trim" as an option for the parameter, because "trim" could implicate that only leading and trailing degenerates are trimmed.